// components/Statistic/StatisticItem.js
import React, { useEffect, useState } from 'react';
import Avatar from '@material-ui/core/Avatar';
import CardHeader from '@material-ui/core/CardHeader';
import Card from '@material-ui/core/Card';
import { makeStyles } from "@material-ui/core/styles";
import { projectDB } from "firebaseConfig/firebaseConfig.js";
import moment from "moment";

const useStyles = makeStyles({
  cardHeader: {
    padding: "8px",
    flexWrap: "wrap",
    flexGrow: 1
  },
  card: {
    borderColor: "rgba(0,0,0,0)",
    minWidth: "100%",
  },
  userPhoto: {
    objectFit: "cover",
    height: "100%",
    width: "100%",
    display: "block"
  }
});


const StatisticItem = ({ timestamp, user }) => {
  const styles = useStyles();
  const [articleOwnerPhoto, setArticleOwnerPhoto] = useState("");
  const [articleOwnerName, setArticleOwnerName] = useState("");

  const getUserNameById = (id) => {
    let name;
    projectDB.ref("users/" + id + "/username").on("value", (snapshot) => {
      name = snapshot.val();
      setArticleOwnerName(name)
    })
  }
  const getUserPhotoById = (id) => {
    let photo;
    projectDB.ref("users/" + id + "/userphoto").on("value", (snapshot) => {
      photo = snapshot.val();
      setArticleOwnerPhoto(photo);
    })
  }

  useEffect(() => {
    getUserPhotoById(user);
    getUserNameById(user);
  }, []);

  return (
    <Card className={styles.card} variant="outlined">
      <CardHeader
        avatar={
          <Avatar>
            {articleOwnerPhoto ? <img className={styles.userPhoto} src={articleOwnerPhoto} alt="" /> : null}
          </Avatar>
        }
        className={styles.cardHeader}
        title={articleOwnerName}
        subheader={moment(new Date(+timestamp)).format("DD.MM.YYYY")}
      />
    </Card >
  );
}

export default StatisticItem;
