// components/TagsBar/SearchTagForm.js
import React, { useRef, useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import LangsTranslate from "LangsTranslate/LangsTranslate.json";
import IconButton from '@material-ui/core/IconButton';
import SearchIcon from '@material-ui/icons/Search';

const useStyles = makeStyles((theme) => ({
  form: {
    width: "100%",
    margin: "0px 0px 10px 0px",
    display: "flex",
    columnGap: "5px",
    alignItems: "center",
    "& > :first-child": {
      flexGrow: 1,
    }
  },
  textField: {
    "& > :first-child": {
      "& > :first-child": {
        padding: "10px"
      }
    }
  }
}));

export default function SearchTagForm({ currentLang, addNewComment }) {
  const classes = useStyles();
  const [commentText, setCommentText] = useState("");
  const [isError, setIsError] = useState(false);

  const tagTextHandler = (e) => {
    if (e.target.value.length > 320) {
      e.preventDefault();
      return;
    }
    if (isError) setIsError(false);
    setCommentText(e.target.value);
  }

  return (
    <form className={classes.form} onSubmit={() => { }} action="#" method="POST">
      <TextField
        label={LangsTranslate[currentLang].tags.searchByTagName}
        className={classes.textField}
        variant="outlined"
        onChange={tagTextHandler}
        value={commentText}
        error={isError}
      />
      <IconButton
        aria-label="send comment"
        color="primary"
        type="submit"
      >
        <SearchIcon />
      </IconButton>
    </form>
  );
}
