// components/TagsBar/TagsBar.js
import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import LangsTranslate from "LangsTranslate/LangsTranslate.json";
import { projectDB } from "firebaseConfig/firebaseConfig.js";
import SearchTagForm from "components/TagsBar/SearchTagForm.js";
import Divider from '@material-ui/core/Divider';

const useStyles = makeStyles({
  root: {
    maxWidth: 275,
    minWidth: 275,
  },
  title: {
    fontSize: 14,
    marginTop: "5px"
  },
  tag: {
    margin: "0",
  },
  cardAction: {
    display: "flex",
    flexWrap: "wrap",
    gap: "5px",
    padding: "0px"
  }
});

export default function TagsBar({ currentLang, isPlanshet }) {
  const classes = useStyles();

  useEffect(() => {
    // projectDB.ref(`tags`).orderByChild('lastUses').limitToLast(10).on('value', (snapshot) => {
    //   const data = snapshot.val();
    //   console.log(data);
    //   setIsEnd(snapshot.numChildren() < articlesAmount);
    //   if (data) setArticles(getArticlesInfo(data));
    //   else setIsNoArticles(true);
    // })
  }, []);

  return (
    <Card style={isPlanshet ? { margin: "20px 0px 0px 0px", maxWidth: "none" } : null} className={classes.root} variant="outlined">
      <CardContent>
        <SearchTagForm currentLang={currentLang} />
        <Divider />
        <Typography className={classes.title} color="textSecondary" gutterBottom>
          {LangsTranslate[currentLang].tags["recentUsedTags"]}
        </Typography>
        <CardActions className={classes.cardAction}>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Dragons</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Happy</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Technology</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">IT</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Clothes</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Learn More</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Learn More</Button>
          <Button style={{ margin: "0" }} variant="contained" color="primary" size="small">Learn More</Button>
        </CardActions>
      </CardContent>
    </Card>
  );
}
