// pages/UserPage/SubsInfo.js
import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import DialogTitle from '@material-ui/core/DialogTitle';
import CancelIcon from '@material-ui/icons/Cancel';
import IconButton from '@material-ui/core/IconButton';
import { makeStyles } from '@material-ui/core/styles';
import StatisticItem from "components/Statistic/StatisticItem.js";
import LangsTranslate from "LangsTranslate/LangsTranslate.json";

const useStyles = makeStyles({
  statisticBody: {
    padding: "5px",
    "& > *": {
      display: "flex",
      flexDirection: "column",
      alignItems: "center"
    }
  },
  icons: {
    padding: '8px 8px 0px 8px',
  },
  iconButtons: {
    borderRadius: "15px",
    border: "1px solid #88888880",
    margin: "0px 5px 0px 0px",
    padding: "7px",
    "& > .MuiButton-label": {
      flexDirection: "column",
    }
  },
});

export default function SubsInfo({ currentLang, subsInfo, subsType }) {
  const styles = useStyles();
  const [open, setOpen] = useState(false);
  const [scroll, setScroll] = useState('paper');
  const userInfoTranslate = LangsTranslate[currentLang].userPage.userInfo;

  const handleClickOpen = (scrollType) => () => {
    setOpen(true);
    setScroll(scrollType);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const descriptionElementRef = React.useRef(null);
  React.useEffect(() => {
    if (open) {
      const { current: descriptionElement } = descriptionElementRef;
      if (descriptionElement !== null) {
        descriptionElement.focus();
      }
    }
  }, [open]);

  return (
    <div>
      <Button
        className={styles.iconButtons}
        onClick={handleClickOpen('paper')}
      >
        <Typography
          variant="body1"
          color="textPrimary"
        >
          {(Object.keys(subsInfo).length).toLocaleString("en")}
        </Typography>
        <Typography
          variant="body2"
          color="textSecondary"
        >
          {userInfoTranslate[subsType]}
        </Typography>
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        scroll={scroll}
        aria-labelledby="scroll-dialog-title"
        aria-describedby="scroll-dialog-description"
      >
        <DialogTitle style={{ textAlign: "center" }} id="scroll-dialog-title">
          {userInfoTranslate[subsType]}
        </DialogTitle>
        <DialogContent className={styles.statisticBody} dividers={scroll === 'paper'}>
          <div>
            {Object.keys(subsInfo).length !== 0 ? Object.entries(subsInfo).map(sub => {
              return <StatisticItem key={sub[0]} timestamp={sub[0]} user={sub[1]} />
            }) :
              <Typography style={{ padding: "12px" }} variant="h6" color="textPrimary">
                {userInfoTranslate[`no${subsType}`]}
              </Typography>}
          </div>
        </DialogContent>
        <DialogActions>
          <IconButton aria-label="cancel" onClick={handleClose} color="primary">
            <CancelIcon />
          </IconButton>
        </DialogActions>
      </Dialog>
    </div >
  );
}
