// pages/UserPage/UserInfo.js
import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import clsx from 'clsx';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Collapse from '@material-ui/core/Collapse';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import SettingsIcon from '@material-ui/icons/Settings';
import TextField from '@material-ui/core/TextField';
import { projectDB } from "firebaseConfig/firebaseConfig.js";
import Loader from "components/Loader/Loader.js";
import LangsTranslate from "LangsTranslate/LangsTranslate.json";

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: "100%",
  },
  expand: {
    transform: 'rotate(0deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  userPhoto: {
    objectFit: "cover",
    height: "100%",
    width: "100%",
    display: "block"
  },
  formItem: {
    minWidth: "100%",
  },
  fieldsetMargins: {
    "& > *": {
      margin: "0px 0px 10px 0px",
    },
    "& > :last-child": {
      margin: 0
    }
  }
}));

export default function UserInfo({ currentLang }) {
  const classes = useStyles();
  const langUserInfo = LangsTranslate[currentLang].userPage.userInfo;
  const [expanded, setExpanded] = React.useState(false);
  const [userInfo, setUserInfo] = useState("");

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };
  const formSubmitHandler = (event) => {
    event.preventDefault();
  }

  const userPhotoHandler = (e) => {
    setUserInfo(prev => (
      { ...prev, photo: e.target.value }
    ))
  }

  React.useEffect(() => {
    if (!localStorage.getItem("currentUserId")) return;
    projectDB.ref('users/' + localStorage.getItem("currentUserId")).on('value', (snapshot) => {
      const data = snapshot.val();
      setUserInfo({ name: data.username, photo: data.userphoto, bio: data.userbio });
    });
  }, []);

  return (
    <Card className={classes.root}>
      <CardHeader
        avatar={
          userInfo ?
            <Avatar aria-label="recipe">
              {userInfo.photo && <img className={classes.userPhoto} src={userInfo.photo} alt="" />}
            </Avatar> :
            <Loader />
        }
        title={userInfo.name}
      />
      {userInfo &&
        <>
          <CardContent style={{ padding: "3px 16px" }}>
            <Typography variant="body2" color="textSecondary" component="p">
              {userInfo.bio}
            </Typography>
          </CardContent>
          <CardActions disableSpacing>
            <Button style={{ marginLeft: "8px" }} variant="outlined" color="primary" aria-label="logout">
              {langUserInfo.logout}
            </Button>
            <IconButton
              className={clsx(classes.expand, {
                [classes.expandOpen]: expanded,
              })}
              onClick={handleExpandClick}
              aria-expanded={expanded}
              aria-label="settings"
            >
              <SettingsIcon />
            </IconButton>
          </CardActions>
        </>}
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <form onSubmit={formSubmitHandler} action="#" method="POST">
            <fieldset className={classes.fieldsetMargins}>
              <fieldset>
                <TextField
                  placeholder={langUserInfo.urlp}
                  className={classes.formItem}
                  label={langUserInfo.urlp}
                  variant="outlined"
                  value={userInfo.photo}
                  onChange={userPhotoHandler}
                />
              </fieldset>
              <fieldset >
                <TextField
                  placeholder={langUserInfo.username}
                  label={langUserInfo.username}
                  className={classes.formItem}
                  variant="outlined"
                  value={userInfo.name}
                // onChange={e => setPassword(e.target.value)}
                />
              </fieldset>
              <fieldset >
                <TextField
                  placeholder={langUserInfo.bio}
                  label={langUserInfo.bio}
                  className={classes.formItem}
                  rows={3}
                  multiline
                  variant="outlined"
                  value={userInfo.bio}
                // onChange={e => setPassword(e.target.value)}
                />
              </fieldset>

              {/* {Exception.isEx && <ErrorMessage />} */}
              <Button
                variant="contained"
                color="primary"
                type="submit"
              >
                {langUserInfo.submit}
              </Button>
            </fieldset>
          </form>
        </CardContent>
      </Collapse>
    </Card>
  );
}
