// pages/Feeds/TagsBar/TagsBar.js
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';

const useStyles = makeStyles({
  root: {
    maxWidth: 275,
    minWidth: 275,
  },
  title: {
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
});

export default function OutlinedCard() {
  const classes = useStyles();

  return (
    <Card className={classes.root} variant="outlined">
      <CardContent>
        <Typography className={classes.title} color="textSecondary" gutterBottom>
          Popular tags
        </Typography>
        <CardActions style={{ display: "flex", flexWrap: "wrap", gap: "5px", padding: "0px" }}>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Dragons</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Happy</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Technolagy</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">IT</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Clothes</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Learn More</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Learn More</Button>
          <Button style={{ margin: "0" }} variant="outlined" size="small">Learn More</Button>
        </CardActions>
      </CardContent>
    </Card>
  );
}
