// pages/NewArticle/NewArticleForm.js
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import FilledInput from '@material-ui/core/FilledInput';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import TextField from '@material-ui/core/TextField';
import InputLabel from '@material-ui/core/InputLabel';
import OutlinedInput from '@material-ui/core/OutlinedInput';

const useStyles = makeStyles((theme) => ({
  root: {
    padding: "20px",
    display: "grid",
    gap: "15px"
  },
  fieldsetStyles: {
    margin: 0,
  }
}));

export default function ComposedTextField() {
  const [name, setName] = React.useState('');
  const styles = useStyles();

  const handleChange = (event) => {
    setName(event.target.value);
  };

  return (
    <form className={styles.root} noValidate autoComplete="off">

      {/* <FormControl>
        <InputLabel htmlFor="component-helper">Name</InputLabel>
        <Input
          id="component-helper"
          value={name}
          onChange={handleChange}
          aria-describedby="component-helper-text"
        />
        <FormHelperText id="component-helper-text">Some important helper text</FormHelperText>
      </FormControl>
      <FormControl error>
        <InputLabel htmlFor="component-error">Name</InputLabel>
        <Input
          id="component-error"
          value={name}
          onChange={handleChange}
          aria-describedby="component-error-text"
        />
        <FormHelperText id="component-error-text">Error</FormHelperText>
      </FormControl> */}
      <fieldset className={styles.fieldsetStyles}>
        <FormControl style={{ width: "100%" }} variant="outlined">
          <InputLabel htmlFor="component-outlined1">Article title</InputLabel>
          <OutlinedInput id="component-outlined1" onChange={handleChange} label="Article title" />
        </FormControl>
      </fieldset>
      <fieldset className={styles.fieldsetStyles}>
        <FormControl style={{ width: "100%" }} variant="outlined">
          <InputLabel htmlFor="component-outlined2">What`s this article about?</InputLabel>
          <OutlinedInput id="component-outlined2" onChange={handleChange} label="What`s this article about?" />
        </FormControl>
      </fieldset>
      <fieldset className={styles.fieldsetStyles}>
        <FormControl style={{ width: "100%" }} variant="outlined">
          <InputLabel htmlFor="component-outlined3">Write your article</InputLabel>
          <OutlinedInput rows={12} multiline id="component-outlined3" value={name} onChange={handleChange} label="Write your article" />
        </FormControl>
      </fieldset>
    </form>
  );
}
