// pages/NewArticle/TextEditor.js
import React, { useState, useMemo, useEffect } from 'react';
import MUIEditor, { MUIEditorState, toHTML } from 'react-mui-draft-wysiwyg';
import textEditorUk from 'pages/NewArticle/textEditorUk.js';

export default React.memo(function TextEditor({ currentLang, setHtml }) {
  const toolbarStyles = {
    style: {
      boxShadow: "none",
      borderBottom: "1px solid #888",
      borderRadius: 0,
      padding: "0px 0px 10px 0px"
    },
  };

  const configEN = {
    lang: 'en',
    toolbar: toolbarStyles,
  }
  const configUK = {
    translations: textEditorUk,
    toolbar: toolbarStyles,
  }

  const convertToHTML = () => {
    const stateHtml = toHTML(editorState.getCurrentContent())
    setHtml(stateHtml)
  }
  const currentConfig = useMemo(() => {
    return currentLang === 'uk' ? configUK : configEN
  }, [currentLang])
  const [editorState, setEditorState] = useState(
    MUIEditorState.createEmpty(currentConfig),
  )
  const onChange = newState => {
    setEditorState(newState);
  }

  useEffect(() => {
    convertToHTML();
  })
  return (
    <MUIEditor
      editorState={editorState}
      onChange={onChange}
      config={currentConfig}
    />
  )
})
