// pages/NewArticle/textEditorUk.js
export default {
  controls: {
    undo: {
      title: 'Шаг назад',
    },
    redo: {
      title: 'Шаг вперед',
    },
    bold: {
      title: 'Жирний текст',
    },
    italic: {
      title: 'Курсивний текст',
    },
    underline: {
      title: 'Пiдкресленний текст',
    },
    strikethrough: {
      title: 'Закреслений текст',
    },
    code: {
      title: 'Код',
    },
    fontColor: {
      title: 'Колір шрифту',
      labels: {
        noColor: 'None',
      },
    },
    fontBackgroundColor: {
      title: 'Колір виділення',
      labels: {
        noColor: 'None',
      },
    },
    linkAdd: {
      title: 'Додати посилання',
      labels: {
        link: 'Посилання',
      },
      actions: {
        add: 'Дабовити',
        cancel: 'Скасувати',
      },
    },
    linkRemove: {
      title: 'Видалити посилання',
    },
    image: {
      title: 'Вставити зображення',
      actions: {
        upload: 'Завантажити зображення',
        url: 'За URL-адресою',
        add: 'Добавити',
        cancel: 'Скасуватти',
        alignLeft: 'Вирівняти ліворуч',
        alignCenter: 'Вирівняти по центру',
        alignRight: 'Вирівняти вправо',
        remove: 'Видалити зображення',
        resize: 'Змінити розмір',
      },
      labels: {
        dropImageHere: 'Опустіть зображення сюди або натисніть, щоб завантажити',
        width: 'Ширина',
        height: 'Висота',
        url: 'URL',
        insertOptions: 'Вставити параметри зображення',
        editOptions: 'Параметри зображення',
      },
      errorMessages: {
        notValidImage: 'Невалідне зображення',
      },
    },
    blockType: {
      title: 'Формат блоку',
      labels: {
        normal: 'Звичайний',
        header1: 'Заголовок 1',
        header2: 'Заголовок 2',
        header3: 'Заголовок 3',
        header4: 'Заголовок 4',
        header5: 'Заголовок 5',
        header6: 'Заголовок 6',
      },
    },
    fontSize: {
      title: 'Розмiр шрифта',
      labels: {
        default: 'за замовчуванням',
      },
    },
    fontFamily: {
      title: 'Тип шрифту',
      labels: {
        default: 'за замовчуванням',
      },
    },
    textAlign: {
      title: 'Вирівняти текст',
      actions: {
        alignLeft: 'Лiворуч',
        alignCenter: 'По центру',
        alignRight: 'Праворуч',
        justify: 'По ширинi',
      },
    },
    unorderedList: {
      title: 'Невпорядкований список',
    },
    orderedList: {
      title: 'Впорядкований список',
    },
  },
};
