// pages/NewArticle/TextEditor.js
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import MUIRichTextEditor from 'mui-rte';
import LangsTranslate from "LangsTranslate/LangsTranslate.json";
import React from 'react';

const defaultTheme = createMuiTheme();

Object.assign(defaultTheme, {
  overrides: {
    MUIRichTextEditor: {
      root: {
        color: "#111",
      },
      editor: {
        background: "#ccc",
        wordWwrap: "break-word",
      },
      toolbar: {
        borderBottom: "1px solid gray",
        background: "rgba(255, 255, 255, 0.1)",
      }
    }
  }
})

const TextEditor = ({ currentLang, defaultValue, articleTextEditorRef, articleSaveHandler }) => {
  const textEditorTextTranslate = LangsTranslate[currentLang].newArticle;

  return (
    <MuiThemeProvider theme={defaultTheme}>
      <MUIRichTextEditor
        controls={[
          "title",
          "bold",
          "italic",
          "underline",
          "strikethrough",
          "highlight",
          "undo",
          "redo",
          "link",
          "media",
          "numberList",
          "bulletList",
          "quote",
          "code"
        ]}
        label={textEditorTextTranslate["article-text-label"]}
        onSave={articleSaveHandler}
        ref={articleTextEditorRef}
        inlineToolbar={true}
        defaultValue={
          defaultValue || JSON.stringify({
            "blocks": [{
              "key": "f3ko9",
              "text": textEditorTextTranslate["article-text"],
              "type": "unstyled",
              "depth": 0,
              "inlineStyleRanges": [],
              "entityRanges": [],
              "data": {}
            }],
            "entityMap": {}
          })
        }
      />
    </MuiThemeProvider>
  )
}

export default TextEditor;
