// components/Comments/CommentNew.js
import React, { useRef, useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import LangsTranslate from "LangsTranslate/LangsTranslate.json";
import IconButton from '@material-ui/core/IconButton';
import SendIcon from '@material-ui/icons/Send';

const useStyles = makeStyles((theme) => ({
  form: {
    width: "100%",
    margin: "10px 0px",
    display: "flex",
    columnGap: "5px",
    alignItems: "center",
    "& > :first-child": {
      flexGrow: 1,
    }
  },
}));

export default function CommentNew({ currentLang, addNewComment }) {
  const classes = useStyles();
  const textFieldRef = useRef();
  const [commentText, setCommentText] = useState("");
  const [isError, setIsError] = useState(false);

  const commentTextHandler = (e) => {
    if (e.target.value.length > 320) {
      e.preventDefault();
      return;
    }
    if (isError) setIsError(false);
    setCommentText(e.target.value);
  }

  useEffect(() => {
    setTimeout(() => {
      if (textFieldRef) {
        textFieldRef.current.focus();
      }
    }, 400)
  }, [])

  return (
    <form className={classes.form} onSubmit={(e) => addNewComment(e, commentText, setIsError, textFieldRef, setCommentText)} action="#" method="POST">
      <TextField
        multiline
        inputRef={textFieldRef}
        id="outlined-basic"
        label={LangsTranslate[currentLang].Feeds.feed["new-comment"]}
        variant="outlined"
        onChange={commentTextHandler}
        value={commentText}
        error={isError}
      />
      <IconButton
        aria-label="send comment"
        color="primary"
        type="submit"
      >
        <SendIcon />
      </IconButton>
    </form>
  );
}
