// components/Comments/Comment.js
import React, { useEffect, useState } from 'react';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import ThumbDownIcon from '@material-ui/icons/ThumbDown';
import ThumbUpIcon from '@material-ui/icons/ThumbUp';
import { makeStyles } from '@material-ui/core/styles';
import { projectDB } from "firebaseConfig/firebaseConfig.js";

const useStyles = makeStyles({
  iconButtons: {
    borderRadius: "15px",
    border: "1px solid #88888880",
    margin: "0px 5px 0px 0px",
    padding: "7px"
  },
  userPhoto: {
    objectFit: "cover",
    height: "100%",
    width: "100%",
    display: "block"
  }
});

export default function RecipeReviewCard({ commentText, dislikes, likes, owner, publishedDate }) {
  const styles = useStyles();
  const [commentLikes, setCommentsLikes] = useState(likes);
  const [commentDislikes, setCommentsDislikes] = useState(dislikes);
  const [articleOwnerPhoto, setArticleOwnerPhoto] = useState("");
  const [articleOwnerName, setArticleOwnerName] = useState("");

  const getUserNameById = (id) => {
    let name;
    projectDB.ref("users/" + id + "/username").on("value", (snapshot) => {
      name = snapshot.val();
    })
    return name;
  }
  const getUserPhotoById = (id) => {
    let photo;
    projectDB.ref("users/" + id + "/userphoto").on("value", (snapshot) => {
      photo = snapshot.val();
    })
    return photo;
  }

  useEffect(() => {
    setArticleOwnerPhoto(getUserPhotoById(owner));
    setArticleOwnerName(getUserNameById(owner));
  }, [])

  return (
    <Card variant="outlined">
      <CardHeader
        avatar={
          <Avatar>
            {articleOwnerPhoto && <img className={styles.userPhoto} src={articleOwnerPhoto} alt="" />}
          </Avatar>
        }
        action={
          <div style={{ margin: "15px 0px 0px 0px", display: "flex", flexWrap: "wrap", rowGap: "5px" }}>
            <IconButton
              aria-label="like"
              className={styles.iconButtons}
              onClick={() => { }}
            >
              <ThumbUpIcon style={{ margin: "0px 5px 0px 0px" }} />
              <span>{commentLikes}</span>
            </IconButton>
            <IconButton
              aria-label="dislike"
              className={styles.iconButtons}
              onClick={() => { }}
            >
              <ThumbDownIcon style={{ margin: "0px 5px 0px 0px" }} />
              <span>{commentDislikes}</span>
            </IconButton>
          </div>
        }
        style={{ flexWrap: "wrap", padding: "15px 15px 8px 15px" }}
        title={articleOwnerName}
        subheader={publishedDate}
      />
      <CardContent style={{ padding: "0px 16px 16px" }}>
        <Typography variant="body1" component="p">
          {commentText}
        </Typography>
      </CardContent>
    </Card>
  );
}
