// components/Comments/CommentNew.js
import React, { useRef, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import LangsTranslate from "LangsTranslate/LangsTranslate.json";
import IconButton from '@material-ui/core/IconButton';
import SendIcon from '@material-ui/icons/Send';

const useStyles = makeStyles((theme) => ({
  form: {
    width: "100%",
    margin: "10px 0px",
    display: "flex",
    columnGap: "5px",
    alignItems: "center",
    "& > :first-child": {
      flexGrow: 1,
    }
  },
}));

export default function CommentNew({ currentLang, addNewComment }) {
  const classes = useStyles();
  const textFieldRef = useRef();

  useEffect(() => {
    setTimeout(() => {
      if (textFieldRef) {
        textFieldRef.current.focus();
      }
    }, 400)
  }, [])

  return (
    <form className={classes.form} autoComplete="off" action="#" method="POST" onSubmit={addNewComment}>
      <TextField multiline inputRef={textFieldRef} id="outlined-basic" label={LangsTranslate[currentLang].Feeds.feed["new-comment"]} variant="outlined" />
      <IconButton
        aria-label="send comment"
        component="span"
        color="primary"
      >
        <SendIcon />
      </IconButton>
    </form>
  );
}
